dali/imgcodec/decoders/tiff_libtiff.cc

```diff
@@ -0,0 +1,225 @@
+    delete helper;
+    return 0;
+  }
+
+ private:
+  std::shared_ptr<InputStream> stream_;
+  ImageSource *in_;
+};
+
+std::unique_ptr<TIFF, void (*)(TIFF *)> OpenTiff(ImageSource *in) {
+  TIFF *tiffptr;
+
+  if (in->Kind() == InputKind::Filename) {
+    tiffptr = TIFFOpen(in->Filename(), "r");
+  } else {
+    TIFFMapFileProc mapproc;
+    if (in->Kind() == InputKind::HostMemory)
+      mapproc = &DecoderHelper::map;
+    else
+      mapproc = nullptr;
+
+    tiffptr = TIFFClientOpen("", "r", reinterpret_cast<thandle_t>(new DecoderHelper(in)),
+                             &DecoderHelper::read,
+                             &DecoderHelper::write,
+                             &DecoderHelper::seek,
+                             &DecoderHelper::close,
+                             &DecoderHelper::size,
+                             mapproc,
+                             /* unmap */ 0);
+  }
+
+  DALI_ENFORCE(tiffptr != nullptr, make_string("Unable to open TIFF image: ", in->SourceInfo()));
+  return {tiffptr, &TIFFClose};
+}
+
+struct TiffInfo {
+  uint32_t image_width, image_height;
+  uint16_t channels;
+
+  uint32_t rows_per_strip;
+  uint16_t bit_depth;
+  uint16_t orientation;
+  uint16_t compression;
+
+  bool is_tiled;
+  bool is_palette;
+};
+
+TiffInfo GetTiffInfo(TIFF *tiffptr) {
+  TiffInfo info = {};
+
+  LIBTIFF_CALL(TIFFGetField(tiffptr, TIFFTAG_IMAGEWIDTH, &info.image_width));
+  LIBTIFF_CALL(TIFFGetField(tiffptr, TIFFTAG_IMAGELENGTH, &info.image_height));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_SAMPLESPERPIXEL, &info.channels));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_BITSPERSAMPLE, &info.bit_depth));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_ORIENTATION, &info.orientation));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_COMPRESSION, &info.compression));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_ROWSPERSTRIP, &info.rows_per_strip));
+
+  info.is_tiled = TIFFIsTiled(tiffptr);
+
+  uint16_t photometric_interpretation;
+  if (TIFFGetField(tiffptr, TIFFTAG_PHOTOMETRIC, &photometric_interpretation) &&
+      photometric_interpretation == PHOTOMETRIC_PALETTE) {
+    info.is_palette = true;
+  }
+
+  return info;
+}
+
+}  // namespace detail
+
+DecodeResult LibTiffDecoderInstance::Decode(SampleView<CPUBackend> out, ImageSource *in,
+                                            DecodeParams opts, const ROI &requested_roi) {
+  auto tiff = detail::OpenTiff(in);
+  auto info = detail::GetTiffInfo(tiff.get());
+
+  // TODO(skarpinski) Support other bit depths
+  DALI_ENFORCE(info.bit_depth == 8, "Not implemented: Only bit depth = 8 is supported");
+
+  // TODO(skarpinski) Support palette images
+  DALI_ENFORCE(!info.is_palette, "Not implemented: Palette images are not yet supported");
+
+  // TODO(skarpinski) Support tiled images
+  DALI_ENFORCE(!info.is_tiled, "Not implemented: Tiled images are not yet supported");
```

Comment:
perhaps we should return `false` in this case, so that we can resort to a different decoder (OpenCV), rather than failing?

Reply:
You're right, fixed